src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -0,0 +1,330 @@
+            # Sort entity_ids to produce the same ordering on multiple runs
+            for i, entity_id in enumerate(sorted(entity_ids)):
+                json_file_name = f'{entity_type}_{i}.json'
+                metadata_file = self.metadata[entity_id]
+                json_content = metadata_file.content
+                metadata[json_file_name] = json_content
+                file_manifest = {
+                    'content-type': 'application/json;',
+                    'crc32c': '0' * 8,
+                    'indexed': True,
+                    'name': json_file_name,
+                    's3_etag': None,
+                    'sha1': None,
+                    'sha256': '0' * 64,
+                    'size': len(json.dumps(json_content)),
+                    'uuid': metadata_file.uuid,
+                    'version': metadata_file.version
+                }
+                manifest.append(file_manifest)
+                if entity_type.endswith('_file'):
+                    file_manifest = self.descriptors[entity_id].to_manifest_entry
+                    manifest.append(file_manifest)
+                else:
+                    pass
+        logger.debug('Bundle contents: %s', list(metadata.keys()))
+        return links_file.version, manifest, metadata
+
+    def _entity_ids_by_type(self,
+                            subgraph_id: str
+                            ) -> Mapping[str, Set[str]]:
+        """
+        Parse the links in a subgraph and return a mapping of the entity types
+        (e.g. 'analysis_file', 'cell_suspension') to a set of entity IDs
+        """
+        links_file: LinksFile = self.links[subgraph_id]
+        links_json = links_file.content
+        entity_ids = defaultdict(set)
+        # Project ID is only mentioned in the links JSON if there is a
+        # supplementary_file_link so add it in here to make sure it is included.
+        entity_ids['project'].add(links_file.project_id)
+        link: JSON
+        for link in links_json['links']:
+            link_type = link['link_type']
+            if link_type == 'process_link':
+                entity_type = link['process_type']
+                entity_id = link['process_id']
+                entity_ids[entity_type].add(entity_id)
+                for category in ('input', 'output', 'protocol'):
+                    for file in link[f'{category}s']:
+                        entity_type = file[f'{category}_type']
+                        entity_id = file[f'{category}_id']
+                        entity_ids[entity_type].add(entity_id)
+            elif link_type == 'supplementary_file_link':
+                for file in link['files']:
+                    entity_type = file['file_type']
+                    entity_id = file['file_id']
+                    entity_ids[entity_type].add(entity_id)
+            else:
+                raise ValueError('Unknown link type', link_type)
+        return dict(entity_ids)
```

Comment:
```suggestion
        entity_ids.default_factory = None
        return entity_ids
```

https://docs.python.org/3/library/collections.html#collections.defaultdict.__missing__

Reply:
Casting the `defaultdict` into a `dict` doesn't appear to be required, so I'll leave it unmodified. I suspect the desire to cast the `defaultdict` arose during incremental development of the file so that its value could be compared against expected.

Reply:
It would be really surprising to be given what the signature claims to be an immutable `Mapping` and then have lookups for missing keys create entries in that mapping. This would be surprising even if the return type was a MutableMapping.

Reply:
Ah I see. Yes, I missed that aspect. Thank you for the explanation.